worker/run.py

```diff
@@ -0,0 +1,399 @@
+        except BundleServiceException:
+            traceback.print_exc()
+        except Exception as e:
+            traceback.print_exc()
+            reply_error(httplib.INTERNAL_SERVER_ERROR, e.message)
+            return
+
+    def write(self, subpath, string):
+        # Make sure you're not trying to write over a dependency.
+        if os.path.normpath(subpath) in self._dep_paths:
+            return
+
+        # Do the write.
+        with open(os.path.join(self._bundle_path, subpath), 'w') as f:
+            f.write(string)
+
+    def kill(self, message):
+        with self._kill_lock:
+            self._killed = True
+            self._kill_message = message
+
+    def _is_killed(self):
+        with self._kill_lock:
+            return self._killed
+
+    def _get_kill_message(self):
+        with self._kill_lock:
+            return self._kill_message
+
+    def _handle_kill(self):
+        if self._is_killed() and not self._docker_container_killed:
+            try:
+                self._docker.kill_container(self._container_id)
+                self._docker_container_killed = True
+            except DockerException:
+                traceback.print_exc()
+
+    def _check_and_report_finished(self):
+        try:
+            finished, exitcode, failure_message = (
+                self._docker.check_finished(self._container_id))
+        except DockerException:
+            traceback.print_exc()
+            return False
+
+        if finished:
+            self._finish(exitcode, failure_message)
+        return finished
+
+    def _finish(self, exitcode=None, failure_message=None):
+        logger.debug('Finished run with UUID %s, exitcode %s, failure_message %s',
+                     self._uuid, exitcode, failure_message)
+        self._set_finished()
+        try:
+            # Delete the container.
+            if self._container_id is not None:
+                while True:
+                    try:
+                        self._docker.delete_container(self._container_id)
+                        break
+                    except DockerException:
+                        traceback.print_exc()
+                        time.sleep(1)
+
+            # Clean-up dependencies.
+            for dep in self._bundle['dependencies']:
+                if not self._worker.shared_file_system:
+                    self._worker.remove_dependee(dep['parent_uuid'], self._uuid)
+                # Docker creates files for each mounted volume. Delete them.
+                child_path = os.path.realpath(
+                    os.path.join(self._bundle_path, dep['child_path']))
+                if child_path.startswith(self._bundle_path):
+                    remove_path(child_path)
+
+            if not self._worker.shared_file_system:
+                logger.debug('Uploading results for run with UUID %s', self._uuid)
+                self._safe_update_run_status('Uploading results')
```

Comment:
We should make a more deliberate stance whether the run_status is going to be more like an enum (small set of values, like state) or an open-ended string.  Using caps and spaces suggests the latter.  Might make sense for it to be the former though (because we have failure_message for details).